### PR TITLE
Add scrolling for bottom-left info panel and logger device for message logs

### LINKED ISSUE #21: `salvage` command
There ought to be a way to retrieve all the inventory and devices from a "dead" robot.  I'm imagining something like `salvage : string -> cmd ()`, where `salvage x` will try to salvage a robot named `x`.  If successful, all of `x`'s inventory and installed devices go into the inventory of the robot that executed `salvage`, and `x` itself disappears.

Some issues to work out:

* What is the range of the `salvage` command?  I'm thinking probably you have to be in the very same cell.
* What if you don't know the name of the robot you want to salvage? There has to be some other way to discover the name(s) of nearby robot(s).  That should be orthogonal to `salvage`, but it's still worth thinking about them together.
* We'll need to add a new `salvage` capability.  What device should confer this capability, and what should its recipe be?  How difficult should it be to construct?

### LINKED ISSUE #36: Wrapping and scrolling for messages
Currently, when displaying the message queue, very simple logic is used to decide how to display the messages and how many to display.  This often leads to messages being unreadable, especially in a smaller terminal.  Also, we choose a static number of messages to display based on the height of the panel, but this is wrong if any of the messages are more than one line high; this leads to messages being cut off the bottom of the panel.

* We ought to use something like `txtWrap` to display the actual messages, so they wrap to whatever width the info panel happens to be.
* Instead of the current logic displaying the last n messages, we should render all the messages into a `brick` `Viewport`.  As a bonus, we should provide some sort of mechanism to allow the user to scroll it.  Note, this kind of scrolling ability will also be needed for entity descriptions, which can  be cut off at the bottom if they are too long and/or the window is too small.

### LINKED ISSUE #39: Add logger device
We should add a new kind of device (a `logger`, recipe: 1 log) which
allows logging messages via a `log` command (i.e. there will be a
`log` capability, provided by a `logger`, and required to execute the
`log` command).  This will also be the mechanism by which uncaught
exceptions are reported; on any robot without a logger, such uncaught
exceptions will just be silently discarded. See this [comment in `Swarm.Game.Step`](https://github.com/byorgey/swarm/blob/main/src/Swarm/Game/Step.hs#L378).

This will require adding a message queue field to the `Robot` record
type.  We should also get rid of the global message queue in the
`GameState`.

We'll have to rethink what the `say` command does---perhaps it should
broadcast a message that will be added to the log of nearby robots?

### LINKED ISSUE #40: Persist entity descriptions in the UI when switching to other panels
Right now, a description of the highlighted inventory entity is shown in the bottom portion of the info panel whenever the info panel is focused.  However, this disappears as soon as you switch away from the info panel.  This feels annoying, though, because an item description might contain something you want to refer to while typing a command at the REPL: some examples of a language feature, for example, or a list of recipes.

The thing that replaces the entity description when not focused on the
info panel is the global message log.  Perhaps the answer is to make
message logs per-robot, and to associate them with a logger device: see #39 .  Then the bottom part of the info
panel will always show info about the currently selected inventory
item, and if you want to see the message log, you can just select that
device in your inventory.

### PR DESCRIPTION
**Pull request by byorgey**
The original main goals of this PR are to (1) solve #36 , that is, make the bottom-left info area fully wrapped and scrollable; and (2) solve #40 , that is, persist the info in that bottom-left info area even when switching to other panels.  In order to address those I ended up making a rather large constellation of related changes, which in the end I hope work well with each other.  Eager to hear others' feedback either from looking at the code or just trying out the game with this PR.

- In order to make the bottom-left area scrollable, split it out into its own panel.  Now we can focus it and then use normal keybindings to scroll it.
- The bottom-left info panel now indicates whether it has additional content scrolled off the top or bottom.
- The bottom-left panel used to display a global list of log/error messages when the inventory was not focused.  So where will those messages be viewable now, if the information displayed in the bottom-left panel persists?  My answer: there is a special `logger` device (#39) which, when focused, now displays the log messages.
    - This means that logs are now per-robot instead of global.
    - The base starts with a `logger` installed.
    - A `log` command can be used to write to a logger.
    - Uncaught exceptions also now write to a logger (instead of using `say`), or are simply discarded if there is no `logger` device installed.
    - The first time a robot's inventory is viewed after a message has been appended to its log, the `logger` device will be automatically focused and the log scrolled all the way to the bottom to show the new entry.  This means that e.g. typing `move` at the REPL will immediately show an exception in the lower left.
- Add a `salvage` command, making it possible to access the log of a dead robot after the fact, once we implement #17 .

There are still a few issues to address which I think can be done in separate PRs:

- The global message log still exists, and the `say` command still appends to it, but there is no way to view it.  Obviously we need to do something with this --- either get rid of it, or figure out a sensible way to view it, etc.
- Exception messages should probably be reformatted to not include any newlines, now that we know they will be wrapped.

Closes #36.  Closes #40.  Closes #39.  Closes #21.

### CODE REVIEW

**Comment by byorgey** [comment]
@xsebek , I know this messes up your distance-measuring functions that used `say`.  Would love to hear your thoughts on that.

File: data/recipes.yaml
Comment (xsebek) [inline data/recipes.yaml:13]: I hope this is a joke otherwise I might need to go cut down a forest or two :fearful: :evergreen_tree: :evergreen_tree: :evergreen_tree:

File: src/Swarm/TUI/Model.hs
Comment (xsebek) [inline src/Swarm/TUI/Model.hs:168]: Nice improvement, the `1` entity installed was a bit confusing :+1:

**Review by xsebek — APPROVED** [review]

**Comment by byorgey** [comment]
> I would just point out, that `install` is hidden under grabber below `--- . . . ----`, so maybe link to that or explicitly add the command here for new players?

Good point.  Updated the `logger` description to try to make this more clear.

> Do you want to remove the say command here, or shall we experiment with the log for a bit and then see if it's a good replacement?

Actually, I was kind of thinking that we might redesign `say` to retain some of its original character, in terms of broadcasting a message.  Like maybe the device needed for `say` is a speaker, and using `say` broadcasts a message to all robots within a certain radius.  (Implementing this might get tricky.)

**Comment by byorgey** [comment]
I've now added a `salvage` command (which, as usual, turned out to be more interesting than I thought --- we didn't already have a way to look up robots by their location, and of course we don't want to have to scan through all robots to see where they are, so...).  This should help with debugging robots that crashed while carrying a `logger`.

I will just update the tutorial tomorrow and then I think this should be ready to merge, if others agree.  But very happy to entertain further questions or suggestions of course.  Thankful for the feedback so far which has made this a lot better already.

**Comment by xsebek** [comment]
@polux well I think players will bother sending robots to `salvage` if they want to get the crashed robots inventory, save the logs for later and clean up. IMHO reasons enough, but we will have to find the right balance when adding antenna #17 and game modes #4.

**Comment by byorgey** [comment]
> how come I can not give away installed devices (e.g. the devices are welded in), but salvage takes them out just fine?

Because salvaging is incredibly destructive, and can only be done by slicing the robot into bits with a high-powered plasma cutter? (Hmm, that gives me an idea for a new device...)

> Also, it currently gives an infinite supply of basic parts

But you already have an infinite supply of basic parts! You cannot do anything different with them than you could do before. This is really a problem with the way `build` works (and will be fixed by #35 ), not with `salvage` per se. 

> Viewing the salvaged robot is weird, fallback to base?

Good catch, I will look into that.

**Comment by xsebek** [comment]
Huh, I can not argue against that reasoning. :slightly_smiling_face:

My point was, that not being able to salvage installed parts would leave the base-on-wheels and infinite-build problems for later. Really we could enable full salvage then, but right now in the "Classic/Creative" modes, it clutters my inventory and is surprisingly stronger than "give" which is provided by grabber too.

**Comment by byorgey** [comment]
> My point was, that not being able to salvage installed parts would leave the base-on-wheels and infinite-build problems for later.

Right, but salvaging installed parts is going to be critical once we implement #35 , I'd rather just have it work the right way from the beginning.  Inventory clutter is definitely a problem that we're going to have to figure out, and it will only get worse once #164 is merged.  In other words just changing the way salvage works is not really going to fix it.  #126 will help, and maybe there can be ways for the user to filter the view, e.g. move some items to a different "I don't care" section at the bottom, toggle whether entities with count 0 are shown, etc.

**Comment by byorgey** [comment]
> surprisingly stronger than "give" which is provided by grabber too.

Yeah, I agree, it is too strong, but we can fix that later.

**Comment by xsebek** [comment]
> how to parse that last sentence

I was a bit worried, that you were planning to fix everything here, glad to hear it is ready :slightly_smiling_face:
Thanks for explaining your plans with relation to other issues, I guess only saw the early game point of view and did not see why it should change :wink:

**Pull request merged**
